src/main/java/mcscheduler/logic/commands/AssignCommand.java

```diff
@@ -91,12 +96,27 @@ public CommandResult execute(Model model) throws CommandException {
                         Messages.MESSAGE_INVALID_ASSIGNMENT_UNAVAILABLE, workerToAssign.getName(), shiftToAssign));
             }
 
+            /*
             if (!shiftToAssign.isRoleRequired(role)) {
                 throw new CommandException(Messages.MESSAGE_INVALID_ASSIGNMENT_NOT_REQUIRED);
             }
+             */
+            // Count for RoleRequirements
+            if (!(role instanceof Leave)) {
+                if (!requiredRoles.containsKey(role)) {
+                    requiredRoles.put(role, getQuantityRequiredForRole(shiftToAssign, role));
+                }
+                requiredRoles.put(role, requiredRoles.get(role) - 1);
+                if (requiredRoles.get(role) < 0) {
+                    // This assignCommand will exceed the role's requirement
+                    throw new CommandException(
+                        String.format(Messages.MESSAGE_INVALID_ASSIGNMENT_NOT_REQUIRED, role, shiftToAssign));
+                }
+            }
```

Comment:
Not sure if this is necessary. Would it work if instead of this, we just put the original check

> if (!shiftToAssign.isRoleRequired(role)) {
>     throw new CommandException(Messages.MESSAGE_INVALID_ASSIGNMENT_NOT_REQUIRED);
> }

within the `for` loop before each `model.addAssignment(assignment)` call (ie. between lines 128 and 129)?

Reply:
Doing it that way will still add the previous assignments before the error assignment to the model. Then the feedback will be a command error rather than a command success, yet it still adds some new assignments. Though if there's a solution to this, do let me know!

Reply:
Since `assignStringBuilder` keeps track of the assignments already added, perhaps we can concat the list of successful assignments to the `CommandException` thrown.

Reply:
I think if it returns an error, you shouldnt accept any part of the command to keep it consistent with other commands. Its also to makes sure the user knows that they're doing something wrong, and not accidentally add unwanted assignments.

Reply:
That's a valid point